org.opent2t.sample.lamp.superpopular/com.wink.lightbulb/js/thingTranslator.js

```diff
@@ -44,107 +51,476 @@ function findResource(schema, di, resourceId) {
 }
 
 /**
- * Wink does not always populate every desired_state property, but last_reading doesn't necessarily
- * update as soon as we send our PUT request. Instead of relying just on one state or the other,
- * we use this StateReader class to read from desired_state if it is there, and fall back to last_reading
- * if it is not.
+ * Validates that an object (arguments[0]) contains properties
+ * (arguments[1...n]) and throws a descriptive OpenT2T error if the
+ * request is malformed.
+ * 
+ * TODO: This method should be moved to a shared location for all translators
  */
-class StateReader {
-    constructor(desired_state, last_reading) {
-        this.desired_state = desired_state;
-        this.last_reading = last_reading;
+function validateHasOwnProperty() {
+    var missing = [];
+    for (var i = 1; i < arguments.length; i++) {
+        if (!arguments[0].hasOwnProperty(arguments[i])) {
+            missing.push(arguments[i]);
+        }
     }
 
-    get(state) {
-        if (this.desired_state[state] !== undefined) {
-            return this.desired_state[state];
-        }
-        else {
-            return this.last_reading[state];
-        }
+    if (missing.length > 0) {
+        throw new OpenT2TError(400, `Request missing ${missing} - ${JSON.stringify(arguments[0])}`);
     }
 }
 
 /**
- * Helper method to convert Wink's 0.0-1.0 brightness scale to a 0-100 scale  
- */ 
-function scaleDeviceBrightnessToTranslatorBrightness(brightnessValue) {
-    return Math.floor(brightnessValue * 100);
+ * Returns a default value if the specified property is null, undefined, or an empty string
+ * 
+ * TODO: This method should be moved to a shared location for all translators
+ */
+function defaultValueIfEmpty(property, defaultValue) {
+    if (property === undefined || property === null || property === "") {
+        return defaultValue;
+    } else {
+        return property;
+    }
 }
 
 /**
- * Helper method to convert a 0-100 scale to Wink's 0.0-1.0 brightness scale
+ * Create a colourRgb resource
+ * 
+ * TODO: This method should be moved to a shared location for all translators
  */
-function scaleTranslatorBrightnessToDeviceBrightness(dimmingValue) {
-    return dimmingValue / 100;
+function createColourRGBResource(expand, rgbArray, range = [0, 255]) {
```

Comment:
The colourRGB resource has a range property, though if it is omitted we should assume [0, 255].  In this case, Wink won't provide it as anything different, but this method is going to be moved to a shared location, and somebody may want to expose a different range depending on the provider.

---
In reply to: [106018057](https://github.com/openT2T/translators/pull/209#discussion_r106018057) [](ancestors = 106018057)